### ISSUE TITLE
Cannot find blosc runtime on macOS 15.0 aarch64

### ISSUE THREAD

**Issue by outpaddling** [PyTables/PyTables#1219]
Trying to build under pkgsrc on my Mac Mini produces the following:

```
* Found bzip2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
* Found blosc headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
.. WARNING:: Could not find the blosc runtime.
   The blosc shared library was *not* found in the default library
   paths. In case of runtime problems, please remember to install it.
* Run 'blosc2_find_directories_hook'
* Unable to find blosc2 wheel.
* Found blosc2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
  * Copying blosc2 runtime library to 'tables' dir because it was not found in standard locations
```

The exact same pkgsrc package works fine on other platforms, e.g. Alma8:

```
* Found bzip2 headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home/bacon/Pkgsrc/pkg/lib``.
* Found blosc headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home/bacon/Pkgsrc/pkg/lib``.
* Run 'blosc2_find_directories_hook'
* Unable to find blosc2 wheel.
* Found blosc2 headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home
```

There are no hacks of any kind in this pkgsrc package: It's using a pristine tables 3.10.1 dist.  I've poked around setup.py a bit, but I'm not seeing anything obvious.  Any hints that help locate the problem would be appreciated.

**Comment by avalentino** [comment]
Dear @outpaddling,
thanks for reporting.
A couple of questions:
1. could you please let us know what command do you use to install PyTables?
Are you using `pip` or `setup.py` directly?
2. apparently `setup.py` finds a `blosc2 ` installation in `/Users/bacon/Pkgsrc/pkg` and `Copying blosc2 runtime library to 'tables' dir`. Could you please confirm that the the library is effectively copied i the `tables` directory?

**Comment by outpaddling** [comment]
> Dear @outpaddling, thanks for reporting. A couple of questions:
> 
>     1. could you please let us know what command do you use to install PyTables?
>        Are you using `pip` or `setup.py` directly?

pkgsrc does not utilize other package managers under the hood, so running setup.py directly.  More extensive output:

===> Building for py312-tables-3.10.1
* Building wheel...
performance hint: tables/utilsextension.pxd:20:26: No exception value declared for 'get_native_type' in pxd file.
Users cimporting this function and calling it without the gil will always require an exception check.
Suggest adding an explicit exception value.
cpuinfo failed, assuming no CPU features: 'flags'
* Using Python 3.12.7 (main, Nov 25 2024, 22:18:45) [Clang 16.0.0 (clang-1600.0.26.3)]
* Found cython 3.0.11
* USE_PKGCONFIG: True
* Found HDF5 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
.. WARNING:: Could not find the HDF5 runtime.
   The HDF5 shared library was *not* found in the default library
   paths. In case of runtime problems, please remember to install it.
* Found LZO 2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
.. WARNING:: Could not find the LZO 2 runtime.
   The LZO 2 shared library was *not* found in the default library
   paths. In case of runtime problems, please remember to install it.
* Skipping detection of LZO 1 since LZO 2 has already been found.
* Found bzip2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
* Found blosc headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
.. WARNING:: Could not find the blosc runtime.
   The blosc shared library was *not* found in the default library
   paths. In case of runtime problems, please remember to install it.
* Run 'blosc2_find_directories_hook'
* Unable to find blosc2 wheel.
* Found blosc2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
  * Copying blosc2 runtime library to 'tables' dir because it was not found in standard locations
Compiling tables/utilsextension.pyx because it changed.
[1/1] Cythonizing tables/utilsextension.pyx
Compiling tables/hdf5extension.pyx because it changed.
[1/1] Cythonizing tables/hdf5extension.pyx
Compiling tables/tableextension.pyx because it changed.
[1/1] Cythonizing tables/tableextension.pyx
Compiling tables/linkextension.pyx because it changed.
[1/1] Cythonizing tables/linkextension.pyx
Compiling tables/_comp_lzo.pyx because it changed.
[1/1] Cythonizing tables/_comp_lzo.pyx
Compiling tables/_comp_bzip2.pyx because it changed.
[1/1] Cythonizing tables/_comp_bzip2.pyx
Compiling tables/lrucacheextension.pyx because it changed.
[1/1] Cythonizing tables/lrucacheextension.pyx
Compiling tables/indexesextension.pyx because it changed.
[1/1] Cythonizing tables/indexesextension.pyx
running bdist_wheel
running build
running build_py
creating build/lib.macosx-15.0-arm64-cpython-312/tables
copying tables/link.py -> build/lib.macosx-15.0-arm64-cpython-312/tables
copying tables/description.py -> build/lib.macosx-15.0-arm64-cpython-312/tables
copying tables/index.py -> build/lib.macosx-15.0-arm64-cpython-312/tables
copying tables/_version.py -> build/lib.macosx-15.0-arm64-cpython-312/tables

> 
>     2. apparently `setup.py` finds a `blosc2 ` installation in `/Users/bacon/Pkgsrc/pkg` and `Copying blosc2 runtime library to 'tables' dir`. Could you please confirm that the the library is effectively copied i the `tables` directory?

Yes, it is copied.  Pkgsrc registers all installed files, and it flagged the extraneous blosc2 lib that was not in the packing list.  This is how I became aware of the problem.

**Comment by avalentino** [comment]
OK, is it a s`.so` or a `.dylib`? A binary file of a symbolic link?
What happens if you put the path into `DYLD_LIBRARY_PATH`?
Did you try to run `otool -L` on the PyTables extensions to understand which version of the library are they linked to?
Sometimes happens that things can go wrong if there are multiple versions of the library in the system.

**Comment by outpaddling** [comment]
> OK, is it a s`.so` or a `.dylib`? A binary file of a symbolic link? 
> Did you try to run `otool -L` on the PyTables extensions to understand which version of the library are they linked to? 

```
Darwin tarpon.acadix  bacon ~/Pkgsrc/pkgsrc/math/py-tables 1016: (pkgsrc): otool -L work/tables-3.10.1/build/lib.macosx-15.0-arm64-cpython-312/tables/tableextension.so
work/tables-3.10.1/build/lib.macosx-15.0-arm64-cpython-312/tables/tableextension.so:
        /Users/bacon/Pkgsrc/pkg/lib/libhdf5.200.dylib (compatibility version 203.0.0, current version 203.0.0)
        /Users/bacon/Pkgsrc/pkg/lib/libblosc.1.dylib (compatibility version 1.0.0, current version 1.21.6)
        /Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib (compatibility version 4.0.0, current version 2.15.1)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)

Darwin tarpon.acadix  bacon ~/Pkgsrc/pkgsrc/math/py-tables 1017: (pkgsrc): ls -l /Users/bacon/Pkgsrc/pkg/lib/libblosc.1.dylib
lrwxr-xr-x  1 bacon  staff  21 Nov 30 07:11 /Users/bacon/Pkgsrc/pkg/lib/libblosc.1.dylib@ -> libblosc.1.21.6.dylib

Darwin tarpon.acadix  bacon ~/Pkgsrc/pkgsrc/math/py-tables 1018: (pkgsrc): ls -l /Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib
lrwxr-xr-x  1 bacon  staff  22 Nov 26 08:17 /Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib@ -> libblosc2.2.15.1.dylib
```

> What happens if you put the path into `DYLD_LIBRARY_PATH`? 

No change.  Pkgsrc is very good at making sure only the correct dependencies are found.  A trick like this will only help in cases where the upstream build system ignores the build env and uses hard-coded paths.  I don't see that sort of thing here.

> Sometimes happens that things can go wrong if there are multiple versions of the library in the system.

The only blosc* installations are in ~/Pkgsrc/pkg.

Thanks...

**Comment by outpaddling** [comment]
Using the following patch:
```
--- setup.py.orig       2024-08-17 08:56:33.000000000 +0000
+++ setup.py
@@ -418,6 +418,7 @@ class BasePackage:
             ]
 
         if use_pkgconfig:
+            print("find_directories() using pkgconfig...")
             # header
             pkgconfig_header_dirs = [
                 Path(d[2:])
@@ -455,6 +456,10 @@ class BasePackage:
 
         hook_dirs = hook() if hook is not None else [None, None, None]
 
+        print("locations = ", locations)
+        print("hook_dirs = ", hook_dirs)
+        print("pkgconfig_dirs = ", pkgconfig_dirs)
+
         directories = [None, None, None]  # headers, libraries, runtime
         for idx, (name, find_path, default_dirs) in enumerate(dirdata):
             use_locations = (
@@ -497,6 +502,8 @@ class BasePackage:
                 else:
                     directories[idx] = Path(path).parent
 
+        print("directories = ", directories)
+        sys.exit(1)
         return tuple(directories)
```

I narrowed down the location of the problem.  macOS and Alma Linux produce different results for the runtime variable:

Mac:
```
===> Building for py312-tables-3.10.1
* Building wheel...
cpuinfo failed, assuming no CPU features: 'flags'
* Using Python 3.12.7 (main, Nov 25 2024, 22:18:45) [Clang 16.0.0 (clang-1600.0.26.3)]
* Found cython 3.0.11
* USE_PKGCONFIG: True
find_directories() using pkgconfig...
locations =  []
hook_dirs =  [None, None, None]
pkgconfig_dirs =  [[], [], []]
directories =  [PosixPath('/Users/bacon/Pkgsrc/pkg/include'), PosixPath('/Users/bacon/Pkgsrc/pkg/lib'), None]
```
Alma 8:
```
===> Building for py312-tables-3.10.1
* Building wheel...
* Using Python 3.12.7 (main, Nov 25 2024, 22:09:48) [GCC 8.5.0 20210514 (Red Hat 8.5.0-22)]
* Found cython 3.0.11
* USE_PKGCONFIG: True
find_directories() using pkgconfig...
locations =  []
hook_dirs =  [None, None, None]
pkgconfig_dirs =  [[], [], []]
directories =  [PosixPath('/home/bacon/Pkgsrc/pkg/include'), PosixPath('/home/bacon/Pkgsrc/pkg/lib'), True]
```

**Comment by avalentino** [comment]
Hi,
I think that the problem is 
```
Darwin tarpon.acadix  bacon ~/Pkgsrc/pkgsrc/math/py-tables 1018: (pkgsrc): ls -l /Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib
lrwxr-xr-x  1 bacon  staff  22 Nov 26 08:17 /Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib@ -> libblosc2.2.15.1.dylib
```
I understand that you have a copy of the symbolic link `libblosc2.4.dylib` but not a copy of the target file `libblosc2.4.dylib`.
Is it correct?

On the other hand `otool` points to the library in `/Users/bacon/Pkgsrc/pkg/lib/libblosc2.4.dylib`, and I assume that there you also have `libblosc2.2.15.1.dylib`.

Probably just removing the dangling link in the PyTables folder could work.
Otherwise you need to copy both the `libblosc.1.dylib@` link and the `libblosc.1.21.6.dylib` library into it.

**Comment by outpaddling** [comment]
Well, the root of the problem is that find_path() is coming up empty on macOS, but returning True on Linux.  How would  libblosc symlinks cause this?

The structure of the blosc installation is exactly that same on the two platforms.  The only difference is "so" vs "dylib".

```
Darwin tarpon.acadix  bacon ~ 1011: (pkgsrc): l Pkgsrc/pkg/lib/libblosc*
-rwxr-xr-x  1 bacon  staff    85K Nov 30 07:11 Pkgsrc/pkg/lib/libblosc.1.21.6.dylib*
lrwxr-xr-x  1 bacon  staff    21B Nov 30 07:11 Pkgsrc/pkg/lib/libblosc.1.dylib@ -> libblosc.1.21.6.dylib
-rw-r--r--  1 bacon  staff    48K Nov 30 07:11 Pkgsrc/pkg/lib/libblosc.a
lrwxr-xr-x  1 bacon  staff    16B Nov 30 07:11 Pkgsrc/pkg/lib/libblosc.dylib@ -> libblosc.1.dylib
-rwxr-xr-x  1 bacon  staff   775K Nov 26 08:17 Pkgsrc/pkg/lib/libblosc2.2.15.1.dylib*
lrwxr-xr-x  1 bacon  staff    22B Nov 26 08:17 Pkgsrc/pkg/lib/libblosc2.4.dylib@ -> libblosc2.2.15.1.dylib
-rw-r--r--  1 bacon  staff   1.1M Nov 26 08:17 Pkgsrc/pkg/lib/libblosc2.a
lrwxr-xr-x  1 bacon  staff    17B Nov 26 08:17 Pkgsrc/pkg/lib/libblosc2.dylib@ -> libblosc2.4.dylib
```
```
Linux alma8.acadix  bacon ~ 1001: (pkgsrc): l Pkgsrc/pkg/lib/libblosc*
-rw-r--r--. 1 bacon bacon  80K Nov 22 15:32 Pkgsrc/pkg/lib/libblosc.a
lrwxrwxrwx. 1 bacon bacon   13 Nov 22 15:32 Pkgsrc/pkg/lib/libblosc.so -> libblosc.so.1*
lrwxrwxrwx. 1 bacon bacon   18 Nov 22 15:32 Pkgsrc/pkg/lib/libblosc.so.1 -> libblosc.so.1.21.6*
-rwxr-xr-x. 1 bacon bacon  55K Nov 22 15:32 Pkgsrc/pkg/lib/libblosc.so.1.21.6*
-rw-r--r--. 1 bacon bacon 1.3M Nov 22 15:32 Pkgsrc/pkg/lib/libblosc2.a
lrwxrwxrwx. 1 bacon bacon   14 Nov 22 15:32 Pkgsrc/pkg/lib/libblosc2.so -> libblosc2.so.4*
-rwxr-xr-x. 1 bacon bacon 804K Nov 22 15:32 Pkgsrc/pkg/lib/libblosc2.so.2.15.1*
lrwxrwxrwx. 1 bacon bacon   19 Nov 22 15:32 Pkgsrc/pkg/lib/libblosc2.so.4 -> lib
```

**Comment by avalentino** [comment]
`find_path()` returning `True` vs `None` is determined by the `find_runtime_path` function.
Since the function uses `ctypes` you could have slightly different behaviours on different OSs

https://github.com/PyTables/PyTables/blob/8f5ad6042a9024e803053bedccaf5c9a105d5133/setup.py#L353-L371

By the way I think that the real problem is not there.
The piece of code copying the libraries apparently cannot work correctly in your case on Darwin:

https://github.com/PyTables/PyTables/blob/8f5ad6042a9024e803053bedccaf5c9a105d5133/setup.py#L944-L955

Probably this is the reason why you have a dangling link in your PyTanles folder at the end.

**Comment by avalentino** [comment]
> I think copying a foreign library into the tables installation should be viewed as a temporary workaround, not a solution.

If you are talking about what is done in the `setup.py` script, please not the it is done specifically in the case in which a wheel is generates.
I totally agree with you, and I don't really like to ship "foreign libraries" in wheels, but AFAIK there is no better solution at th e moment. From my side I personally discourage the use of wheels for PyTables in favor of the use of general purpose package managers lie apt/rpm, conda, etc.

Regarding the specific point of `find_runtime_path()`, my understanding is that, in your setup, libraries are installed in a directory tree having `/Users/bacon/Pkgsrc/pkg` as prefix.
Being it not a standard location I'm not surprised that `find_runtime_path()` doesn't find it, unless you use (DY)LD_LIBRARY_PATH or any kind of custom configuration of the system linker.

The surprising thing, for me, is that the runtime us found in Alma 8 case.

Please note that `find_runtime_path()` basically consists in calling  `ctypes.CDLL(f"{prefix}{self.runtime_name}{suffix}")` with different lib suffixes (while `prefix` is something in`_library_prefixes = ["lib"]`).

Do I miss something?

**Comment by esaaprillia** [comment]
> Trying to build under pkgsrc on my Mac Mini produces the following:
> 
> ```
> * Found bzip2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
> * Found blosc headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
> .. WARNING:: Could not find the blosc runtime.
>    The blosc shared library was *not* found in the default library
>    paths. In case of runtime problems, please remember to install it.
> * Run 'blosc2_find_directories_hook'
> * Unable to find blosc2 wheel.
> * Found blosc2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
>   * Copying blosc2 runtime library to 'tables' dir because it was not found in standard locations
> ```
> 
> The exact same pkgsrc package works fine on other platforms, e.g. Alma8:
> 
> ```
> * Found bzip2 headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home/bacon/Pkgsrc/pkg/lib``.
> * Found blosc headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home/bacon/Pkgsrc/pkg/lib``.
> * Run 'blosc2_find_directories_hook'
> * Unable to find blosc2 wheel.
> * Found blosc2 headers at ``/home/bacon/Pkgsrc/pkg/include``, library at ``/home
> ```
> 
> There are no hacks of any kind in this pkgsrc package: It's using a pristine tables 3.10.1 dist. I've poked around setup.py a bit, but I'm not seeing anything obvious. Any hints that help locate the problem would be appreciated.



@outpaddling 


does this warning stop you from building pytables?

I tried to cross-compile in OpenWRT the warning was almost the same but the cross-compilation process continued and was successful in building

```
2024-12-05T12:59:23.0728855Z * Getting build dependencies for wheel...                                                                                                                                                                                                                                                                                 2024-12-05T12:59:27.1508442Z performance hint: tables/utilsextension.pxd:20:26: No exception value declared for 'get_native_type' in pxd file.                                                                                                                                                                                                         2024-12-05T12:59:27.1510005Z Users cimporting this function and calling it without the gil will always require an exception check.                                                                                                                                                                                                                     2024-12-05T12:59:27.1510961Z Suggest adding an explicit exception value.                                                                                                                                                                                                                                                                               2024-12-05T12:59:27.3621454Z * Using Python 3.11.7 (main, Nov 29 2024, 17:25:36) [GCC 11.4.0]                                                                                                                                                                                                                                                          2024-12-05T12:59:27.3622153Z * Found cython 3.0.10                                                                                                                                                                                                                                                                                                     2024-12-05T12:59:27.3622562Z * USE_PKGCONFIG: True
2024-12-05T12:59:27.3624207Z * Found HDF5 headers at ``/home/staging_dir/target-aarch64_generic_musl/usr/include``, library at ``/home/staging_dir/target-aarch64_generic_musl/usr/lib``.                                                                                                  2024-12-05T12:59:27.3626607Z .. WARNING:: Could not find the HDF5 runtime.                                                                                                                                                                                                                                                                             2024-12-05T12:59:27.3627038Z    The HDF5 shared library was *not* found in the default library
2024-12-05T12:59:27.3627533Z    paths. In case of runtime problems, please remember to install it.
2024-12-05T12:59:27.3628677Z * Found LZO 2 headers at ``/home/staging_dir/target-aarch64_generic_musl/usr/include``, library at ``/home/staging_dir/target-aarch64_generic_musl/usr/lib``.
2024-12-05T12:59:27.3629712Z .. WARNING:: Could not find the LZO 2 runtime.
2024-12-05T12:59:27.3630105Z    The LZO 2 shared library was *not* found in the default library
2024-12-05T12:59:27.3630579Z    paths. In case of runtime problems, please remember to install it.
2024-12-05T12:59:27.3631062Z * Skipping detection of LZO 1 since LZO 2 has already been found.
2024-12-05T12:59:27.3632180Z * Found bzip2 headers at ``/home/staging_dir/target-aarch64_generic_musl/usr/include``, library at ``/home/staging_dir/target-aarch64_generic_musl/usr/lib``.
2024-12-05T12:59:27.3633278Z * pkg-config header dirs for blosc: /home/staging_dir/target-aarch64_generic_musl/usr/include
2024-12-05T12:59:27.3634045Z * pkg-config library dirs for blosc: /home/staging_dir/target-aarch64_generic_musl/usr/lib
2024-12-05T12:59:27.3635114Z * Found blosc headers at ``/home/staging_dir/target-aarch64_generic_musl/usr/include``, library at ``/home/staging_dir/target-aarch64_generic_musl/usr/lib``.
2024-12-05T12:59:27.3635958Z .. WARNING:: Could not find the blosc runtime.
2024-12-05T12:59:27.3636283Z    The blosc shared library was *not* found in the default library
2024-12-05T12:59:27.3636666Z    paths. In case of runtime problems, please remember to install it.
2024-12-05T12:59:27.3637264Z * pkg-config header dirs for blosc2: /home/staging_dir/target-aarch64_generic_musl/usr/include
2024-12-05T12:59:27.3638019Z * pkg-config library dirs for blosc2: /home/staging_dir/target-aarch64_generic_musl/usr/lib
2024-12-05T12:59:27.3638730Z * Run 'blosc2_find_directories_hook'
2024-12-05T12:59:27.3639043Z * Unable to find blosc2 wheel.
2024-12-05T12:59:27.3639834Z * Found blosc2 headers at ``/home/staging_dir/target-aarch64_generic_musl/usr/include``, library at ``/home/staging_dir/target-aarch64_generic_musl/usr/lib``.
2024-12-05T12:59:27.3640805Z   * Copying blosc2 runtime library to 'tables' dir because it was not found in standard locations
```

```
root@OpenWrt:~/mosdns# python -c "import numpy; print(numpy.__version__)"
2.1.3
root@OpenWrt:~/mosdns# python -c"import tables; tables.print_versions()"
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.10.1
HDF5 version:        1.14.4-2
NumPy version:       2.1.3
Numexpr version:     2.10.2 (not using Intel's VML/MKL)
Zlib version:        1.3.1 (in Python interpreter)
LZO version:         2.10 (Mar 01 2017)
BZIP2 version:       1.0.8 (13-Jul-2019)
Blosc version:       1.21.6 (2024-06-24)
Blosc compressors:   blosclz (2.5.1), lz4 (1.10.0), lz4hc (1.10.0), zlib (1.3.1), zstd (1.5.6)
Blosc filters:       shuffle, bitshuffle
Blosc2 version:      2.15.2 (2024-11-28)
Blosc2 compressors:  blosclz (2.5.3), lz4 (1.10.0), lz4hc (1.10.0), zlib (1.3.1), zstd (1.5.6)
Blosc2 filters:      shuffle, bitshuffle
Python version:      3.11.7 (main, Dec  3 2024, 21:24:36) [GCC 13.3.0]
Platform:            Linux-6.6.63-aarch64-with-glibc2.0
Byte-ordering:       little
Detected cores:      4
Default encoding:    utf-8
Default FS encoding: utf-8
Default locale:      (None, None)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
```

**Comment by outpaddling** [comment]
It also works fine on NetBSD and SmartOS at the least.  It's just macOS that's flopping.

I'll try diving a little deeper into find_runtime_path() and see why it's behaving differently on macOS.  I'm not much of a pythonista, but I can read the code with a little patience and research.  In my view, a proper build system looks where you tell it to, not just in fixed standard directories.  This does seem to be happening on all platforms except macOS.  Pkgsrc is very good at communicating the location of dependencies using standard variables and flags (e.g. -L, -rpath in the case of compiled languages).

In the meantime, I committed a hack to set the runtime dir to True when it should be.  This unbreaks the pkgsrc build for now, but of course should be replaced with a root cause fix ASAP.

```
+                print("path is not set!!  This is a problem.")
+                if name == "blosc" or name == "blosc2":
+                    print("directories = ", directories)
+                    print("rundir (should be 'True') = ", directories[2])
+                    # FIXME: tables-3.10.1 hack
+                    # Forcing rundir to True for macOS.  This is a
+                    # hack, but it's the value it should have, since the
+                    # headers and libs are found.  When rundir is None,
+                    # setup.py copyies libblosc into the py-tables
+                    # installation, breaking PLIST.
+                    if directories[0] and directories[1]:
+                        directories[2] = True
```
Confirmed that the latest bulk build worked (absence of py-tables in the following list):

[https://mail-index.netbsd.org/pkgsrc-bulk/2024/12/03/msg026194.html](https://mail-index.netbsd.org/pkgsrc-bulk/2024/12/03/msg026194.html)

**Comment by outpaddling** [comment]
> Copying blosc2 runtime library to 'tables'

This is the problem for pkgsrc.  It flags the foreign file in the packing list as an error.  Only happens on macOS.

**Comment by avalentino** [comment]
> In my view, a proper build system looks where you tell it to, not just in fixed standard directories. 

To be clear, I think that the current status of the `setup.py` script is "not ideal" (to be very generous), but I don't want to discuss the reasons why it is in the current state now.

By the way the `setup.py` script does not just look at the standard locations.

> Pkgsrc is very good at communicating the location of dependencies using standard variables and flags (e.g. -L, -rpath in the case of compiled languages).

We use CFLAGS, LDFLAGS etc. but we do not parse the compiler/linker flags so we have no way to understand if the libraries are in a place that can be found at runtime.

I have to say that I maintain also the debian packages for PyTables, and I completely understand your "packages" perspective.

Maybe we just need a kind of `--packaging-mode` CLI flag preventing the script to be too invasive.

**Comment by esaaprillia** [comment]
@avalentino 

hdf5 problem is also here. it even fails when cross-compiling h5py

**Comment by outpaddling** [comment]
> > What happens if you put the path into `DYLD_LIBRARY_PATH`?
> 
> No change. Pkgsrc is very good at making sure only the correct dependencies are found. A trick like this will only help in cases where the upstream build system ignores the build env and uses hard-coded paths. I don't see that sort of thing here.

Hmm, maybe I had a typo in my previous test?  I tried this again and find_runtime_path() now returns True, as it should.

I'll dig into the pkgsrc infrastructure and see if it's feasible to add a Mac-specific tweak to locate the dylibs.  The standard BSD/Linux methods are clearly not working on Mac.

**Comment by avalentino** [comment]
Good
Please let us know

**Comment by outpaddling** [comment]
I noticed that find_runtime_path() is receiving locations as an argument, but not using it.  If we utilize it to construct  absolute pathnames for CDLL, the function works, and can return an actual path instead of just True:

```
         # dlopen() won't tell us where the file is, just whether
         # success occurred, so this returns True instead of a filename
-        for prefix in self._runtime_prefixes:
-            for suffix in self._runtime_suffixes:
-                try:
-                    ctypes.CDLL(f"{prefix}{self.runtime_name}{suffix}")
-                except OSError:
-                    pass
-                else:
-                    return True
+        for location in locations:
+            for prefix in self._runtime_prefixes:
+                for suffix in self._runtime_suffixes:
+                    path = f"{location}/{prefix}{self.runtime_name}{suffix}"
+                    print("find_runtime_path() trying ", path)
+                          
+                    try:
+                        ctypes.CDLL(path)
+                    except OSError:
+                        pass
+                    else:
+                        # Return path here instead?
+                        return True
```

Output:

```
use_locations =  [PosixPath('/Users/bacon/Pkgsrc/pkg/lib'), PosixPath('/usr/local/lib64'), PosixPath('/usr/local/lib'), PosixPath('/sw/lib64'), PosixPath('/sw/lib'), PosixPath('/opt/lib64'), PosixPath('/opt/lib'), PosixPath('/opt/local/lib64'), PosixPath('/opt/local/lib'), PosixPath('/usr/lib64'), PosixPath('/usr/lib'), PosixPath('/lib64'), PosixPath('/lib')]
find_runtime_path() trying  /Users/bacon/Pkgsrc/pkg/lib/libblosc2.so
find_runtime_path() trying  /Users/bacon/Pkgsrc/pkg/lib/libblosc2.dylib
path =  /Users/bacon/Pkgsrc/pkg/lib/libblosc2.dylib
* Found blosc2 headers at ``/Users/bacon/Pkgsrc/pkg/include``, library at ``/Users/bacon/Pkgsrc/pkg/lib``.
```

**Comment by avalentino** [comment]
@outpaddling good solution, but I would like to keep also the version without `locations` in the search sequence to give a chance to the system configuration.
Do you want to submit a PR?

**Comment by outpaddling** [comment]
> @outpaddling good solution, but I would like to keep also the version without `locations` in the search sequence to give a chance to the system configuration. Do you want to submit a PR?

Done.

**Comment by avalentino** [comment]
Closed via #1223.
